src/gfx.cpp

```diff
@@ -1775,49 +1791,62 @@ static void SplitDirtyBlocks(uint start, int left, int top, int right, int botto
-		Rect &r = _dirty_blocks[start];
-		if (left > r.right || right < r.left || top > r.bottom || bottom < r.top) {
+		/* Iterate through the dirty blocks. It is not valid to use a ranged for loop here, because
+		 * the vector holding them might be reallocated during iteration, and that would invalidate
+		 * the iterator.
+		 */
+		Rect &current = _dirty_blocks[start];
+		if (left > current.right || right < current.left || top > current.bottom || bottom < current.top) {
+			/* No overlap. */
 			continue;
 		}
-		/* overlap or contact of some sort */
-		if (left >= r.left && right <= r.right && top >= r.top && bottom <= r.bottom) {
-			/* entirely contained by existing  */
+		if (left >= current.left && right <= current.right && top >= current.top && bottom <= current.bottom) {
+			/* New rectangle is entirely contained by existing rectangle. */
 			return;
 		}
-		if (left <= r.left && right >= r.right && top <= r.top && bottom >= r.bottom) {
-			/* entirely contains existing */
-			r = _dirty_blocks.back();
+		if (left <= current.left && right >= current.right && top <= current.top && bottom >= current.bottom) {
+			/* New rectangle entirely contains existing rectangle. */
+			current = _dirty_blocks.back();
 			_dirty_blocks.pop_back();
 			start--;
 			continue;
 		}
-		if (left < r.left && right > r.left) {
-			int middle = r.left;
+		if (left < current.left && right > current.left) {
+			/* Partial overlap, split at left edge. Because SplitDirtyBlocks calls `push_back`,
+			 * which might invalidate the reference to the current rectangle by reallocating the
+			 * vector somewhere else, it is important to cache the boundary in a temporary.
+			 */
+			int middle = current.left;
 			SplitDirtyBlocks(start, left, top, middle, bottom);
 			SplitDirtyBlocks(start, middle, top, right, bottom);
 			return;
 		}
-		if (right > r.right && left < r.right) {
-			int middle = r.right;
+		if (right > current.right && left < current.right) {
+			/* Partial overlap, split at right edge. */
+			int middle = current.right;
 			SplitDirtyBlocks(start, left, top, middle, bottom);
 			SplitDirtyBlocks(start, middle, top, right, bottom);
 			return;
 		}
-
-		if (top < r.top && bottom > r.top) {
-			int middle = r.top;
+		if (top < current.top && bottom > current.top) {
+			/* Partial overlap, split at top edge. */
+			int middle = current.top;
 			SplitDirtyBlocks(start, left, top, right, middle);
 			SplitDirtyBlocks(start, left, middle, right, bottom);
 			return;
 		}
-
-		if (bottom > r.bottom && top < r.bottom) {
-			int middle = r.bottom;
+		if (bottom > current.bottom && top < current.bottom) {
+			/* Partial overlap, split at bottom edge. */
+			int middle = current.bottom;
 			SplitDirtyBlocks(start, left, top, right, middle);
 			SplitDirtyBlocks(start, left, middle, right, bottom);
 			return;
 		}
+
+		/* New rectangle shares exactly one side with existing rectangle; possible merging opportunity. */
```

Comment:
This point in the code is reached iff the new rectangle shares exactly one common edge with the current rectangle; I have painstakingly worked through the boolean algebra to prove this is the case.

This is, therefore, where one would insert code to merge with an existing rectangle, if there is a tangible benefit to doing so.

Reply:
Comment should probably be expanded to reflect this then :)

Reply:
> Comment should probably be expanded to reflect this then :)

Good idea. I thought about leaving the boolean algebra in place as a comment; but it made the function about 100 lines longer.